titus-client/src/main/java/com/netflix/titus/runtime/endpoint/v3/grpc/ProtobufCache.java

```diff
@@ -0,0 +1,151 @@
+import com.netflix.spectator.api.Registry;
+import com.netflix.spectator.api.patterns.PolledMeter;
+import com.netflix.titus.common.framework.scheduler.ScheduleReference;
+import com.netflix.titus.common.framework.scheduler.model.ScheduleDescriptor;
+import com.netflix.titus.common.runtime.TitusRuntime;
+import com.netflix.titus.common.util.CollectionsExt;
+import com.netflix.titus.common.util.time.Clock;
+
+class ProtobufCache<E, C> {
+
+    private static final String METRICS_ROOT = "protobufCache.";
+
+    private final Function<E, C> converter;
+    private final Supplier<Predicate<E>> expiredPredicateSupplier;
+    private final GrpcObjectsCacheConfiguration configuration;
+    private final Clock clock;
+
+    @VisibleForTesting
+    final ConcurrentMap<String, CacheEntry<E, C>> cache = new ConcurrentHashMap<>();
+
+    @VisibleForTesting
+    final ConcurrentMap<String, Long> toBeRemoved = new ConcurrentHashMap<>();
+
+    private final ScheduleReference scheduleRef;
+
+    private final Counter cacheHit;
+    private final Counter cacheMiss;
+
+    ProtobufCache(String name,
+                  Function<E, C> converter,
+                  Supplier<Predicate<E>> expiredPredicateSupplier,
+                  GrpcObjectsCacheConfiguration configuration,
+                  TitusRuntime titusRuntime) {
+        this.converter = converter;
+        this.expiredPredicateSupplier = expiredPredicateSupplier;
+        this.configuration = configuration;
+        this.clock = titusRuntime.getClock();
+
+        Registry registry = titusRuntime.getRegistry();
+        PolledMeter.using(registry).withId(registry.createId(METRICS_ROOT + "cacheSize", "cache", name)).monitorSize(cache);
+        PolledMeter.using(registry).withId(registry.createId(METRICS_ROOT + "toBeRemoved", "cache", name)).monitorSize(toBeRemoved);
+        this.cacheHit = registry.counter(METRICS_ROOT + "cacheAccess", "cache", name, "cacheHit", "true");
+        this.cacheMiss = registry.counter(METRICS_ROOT + "cacheAccess", "cache", name, "cacheHit", "false");
+
+        ScheduleDescriptor gcScheduleDescriptor = ScheduleDescriptor.newBuilder()
+                .withName(ProtobufCache.class.getSimpleName() + "#" + name)
+                .withDescription("Protobuf entity cache")
+                .withInitialDelay(Duration.ofMillis(configuration.getCacheRefreshInitialDelayMs()))
+                .withInterval(Duration.ofMillis(configuration.getCacheRefreshIntervalMs()))
+                .withTimeout(Duration.ofMillis(configuration.getCacheRefreshTimeoutMs()))
+                .build();
+        this.scheduleRef = titusRuntime.getLocalScheduler().schedule(gcScheduleDescriptor, context -> removeExpiredEntries(), true);
+    }
+
+    void shutdown() {
+        scheduleRef.cancel();
+    }
+
+    C get(String id, E entity) {
+        CacheEntry<E, C> entry = cache.get(id);
+        if (entry != null) {
+            if (entry.getCoreEntity() == entity) {
```

Comment:
Do you think we should use equality check here? I am curious how many times we would hit the negative of this condition but not a complete cache miss.

Reply:
No, it would be too expensive. It is done on purpose like that. JobService will always return the same, latest value. If that value changes (even if equals is true), we should update the cache so we can depend on fast reference comparison.